### ISSUE TITLE
[jvm-packages] XGBoostClassificationModel prediction results are wrong.

### ISSUE THREAD

**Issue by sethah** [dmlc/xgboost#2840]
When using `XGBoostClassificationModel` that was trained with either `multi:softmax` or `binary:logitraw` objectives, the predictions will not make sense. This is because the code assumes that the the output of the booster predict function are probabilities, when sometimes they are not. This assumption is hardcoded in `predictRaw`:

````scala
val transformerForProbabilitiesArray =
      (rawPredArray: mutable.WrappedArray[Float]) => 
        if (numClasses == 2) {
          Array(1 - rawPredArray(0), rawPredArray(0)).map(_.toDouble)
        } else {
          rawPredArray.map(_.toDouble).array
        }
````

The simplest fix for this would probably be just to disallow those objective functions to be used with `XGBoostClassificationModel` - however, that is problematic because (as far as I can tell) the `XGBoostClassificationModel` object has no way of knowing what objective function was used.

IMO, a more comprehensive solution is warranted. In particular, I think the following class structure works better:

````scala
class XGBoostRegressor extends Predictor[Vector, XGBoostRegressor, XGBoostRegressionModel]
class XGBoostRegressionModel extends PredictionModel[Vector, XGBoostRegressionModel]

class XGBoostClassififier extends ProbabilisticClassifier[Vector, XGBoostClassifier, XGBoostClassificationModel]
class XGBoostClassificationModel extends ProbabilisticClassificationModel[Vector, XGBoostClassificationModel]
````

This is the same structure as the Spark `DecisionTreeClassifier/Regressor` for example. The problem I run into with this solution is that, for the transform method in the classifier, you would typically grab the raw prediction margins from the `Booster` object, and then transform the raw margins to probabilities and further to predictions, but there isn't a single method in the C API that can do that (AFAIK) and so you'd either need to make two passes over the data, or you'd need to have some way of converting raw margins to probabilities for each specific objective function. This is possible but perhaps not ideal.

Re-structuring the estimators and transformers for XGBoost this way would make the API more usable, and prevent users from manually casting to either regression or classification model (e.g. `val model = estimator.train(df).asInstanceOf[XGBoostClassificationModel]`. If I've missed or misunderstood something please let me know.

Also, I understand that the design of the API is a slightly separate issue (though I do think it would correct the issue with incorrect predictions), so if these concerns should be separated to different issues then I can do that.

## Environment info
Operating System: Mac OS

Package used (python/R/jvm/C++): jvm

`xgboost` version used: Source

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`) ac7a9edb06f2641bc68ba8edd4ce241259e20cbc

## Steps to reproduce

````scala
    val numIterations = 5
    val paramMap = Map("eta" -> "1", "max_depth" -> "6", "silent" -> "1",
      "objective" -> "multi:softmax", "num_class" -> "6")
    val trainingDF = buildDataFrame(MultiClassification.train)
    val xgBoostModelWithDF = XGBoost.trainWithDataFrame(trainingDF, paramMap,
      round = numIterations, nWorkers = numWorkers)
    val predResultsFromDF = xgBoostModelWithDF.transform(trainingDF)
    predResultsFromDF.show()
````

```
+---+-----+--------------------+-------------+----------+
| id|label|            features|probabilities|prediction|
+---+-----+--------------------+-------------+----------+
|  0|  1.0|[2.0,2.0,0.0,3.0,...|        [1.0]|       0.0|
|  1|  0.0|[3.0,3.0,3.0,2.0,...|        [0.0]|       0.0|
|  2|  2.0|[2.0,1.0,2.0,3.0,...|        [2.0]|       0.0|
|  3|  0.0|[2.0,2.0,2.0,0.0,...|        [0.0]|       0.0|
+---+-----+--------------------+-------------+----------+

```

**Comment by sethah** [comment]
cc @superbobry @CodingCat

**Comment by superbobry** [comment]
@sethah I am using xgboost4j only for learning the model, so I've never encountered the issue you describe, however, the refactoring you suggest looks reasonable 👍 .
